syntax = "proto3";

package nh;

message Login {
  int32 player_id = 1;
  int32 session_start_time = 2;
}

message LoginStatus {
  bool success = 1;
  int32 player_id = 2;
}

message Player { string username = 1; }

message RequestClan {}

message Clan { repeated Player players = 1; }

message BagInventory {}

message InsertItem { BagItem item = 1; }

message RetrieveItem { BagItem item = 1; }

message RetrieveItemStatus {
  BagItem item = 1;
  bool success = 2;
}

message SaveEquipment { Equipment equipment = 1; }

message Equipment {
  int32 slot = 1;
  bytes item = 2;
}

message SavedEquipment {
  repeated Equipment equipments = 1;
}
message RetrieveSavedEquipment { }

message BagItem {
  int32 id = 1;
  bytes item = 2;
}

message Bag { repeated BagItem items = 1; }

message Objective {
  string name = 1;
  int32 reward = 2;
}

message CompleteTask { string objective_name = 1; }

message OpenLootbox { int32 rarity = 1; }

message ClanPower { 
  string name = 1;
  int32 num = 2;
}

message RetrieveClanPowers { }

message ClanPowers {
  repeated ClanPower powers = 1;
}

message Reward {
  int32 reward = 1;
  string objective = 2;
  int32 total_reward = 3;
}

message Status {
  int32 code = 1;
  string error_message = 2;
}

message SessionEvent {
  int32 session_turn = 1;
  string name = 2;
  int32 previous_value = 3;
  int32 value = 4;
  string string_value = 5;
}

message Event {
    oneof msg {
        RequestClan request_clan = 1; 
        Clan clan = 2; 
        Player player = 3;
        Login login = 4;
        BagInventory bag_inventory = 5;
        BagItem bag_item = 6;
        Bag bag = 7;
        InsertItem insert_item = 8;
        RetrieveItem retrieve_item = 9;
        CompleteTask complete_task = 10;
        Reward reward = 11;
        OpenLootbox open_lootbox = 12;
        RetrieveClanPowers clan_powers = 13;
        RetrieveSavedEquipment retrieve_saved_equipment = 14;
        SaveEquipment save_equipment = 15;
        SessionEvent session_event = 16;
    }
}
